recipe/patches/0015-Add-CondaEcosystemModifyDllSearchPath.patch

```diff
@@ -1,7 +1,48 @@
-From 9eb3eea71dcbe3c9c2649169bd35614998b1c650 Mon Sep 17 00:00:00 2001
+From 5bb1915898462b89ea9ed9791e26a2e7e3c63fd2 Mon Sep 17 00:00:00 2001
 From: Ray Donnelly <mingw.android@gmail.com>
 Date: Tue, 24 Dec 2019 18:37:17 +0100
-Subject: [PATCH 20/30] Add CondaEcosystemModifyDllSearchPath()
+Subject: [PATCH] Add CondaEcosystemModifyDllSearchPath()
+
+There are 2 modes depending on CONDA_DLL_SEARCH_MODIFICATION env variable
+
+- unset CONDA_DLL_SEARCH_MODIFICATION (Default)
+
+  In this mode, the python interpreter works as if the python interpreter
+  was called with the following directories.
+
+    os.add_dll_directory(join(sys.prefix, 'bin'))
+    os.add_dll_directory(join(sys.prefix, 'Scripts'))
+    os.add_dll_directory(join(sys.prefix, 'Library', 'bin'))
+    os.add_dll_directory(join(sys.prefix, 'Library', 'usr', 'bin'))
+    os.add_dll_directory(join(sys.prefix, 'Library', 'mingw-w64', 'bin'))
+
+  Search order
+    - The directory that contains the DLL
+    - Application directory
+    - Directories added with os.add_dll_directory
+    - Above 5 directories
+    - C:\Windows\System32
+
+  Note that the default behaviour changed in conda python 3.9 to
+  make os.add_dll_directory work in user code.
+
+- CONDA_DLL_SEARCH_MODIFICATION=1
+
+  Search order is roughly,
+
+    - The directory that contains the DLL
+    - Application directory
+    - C:\Windows
+    - Current working directory
+    - Above 5 directories if not in PATH
+    - PATH
+    - Directories added with os.add_dll_directory
+    - Old PATH entries (Deficiency in current patch)
+    - Old working directories (Deficiency in current patch)
+    - C:\Windows\System32
```

Comment:
@conda-forge/python, I've kept the `CONDA_DLL_SEARCH_MODIFICATION=1` code path as before in 3.8 and 3.9.
`CONDA_DLL_SEARCH_MODIFICATION` unset is changed now from 3.8 and 3.9.

In python3.8 upstream made the decision to change the DLL search order so that PATH is not looked at and instead asking library authors to rely on `os.add_dll_directory`. This broke conda and we reverted the upstream work. This came with the caveat that `os.add_dll_directory` doesn't work in conda python. See https://github.com/conda/conda/issues/10897. Although I wasn't in the conversation to revert upstream work to support I think it made sense at the time as otherwise it would have meant changing hundreds of packages.

I think with python3.10, we should go with upstream decision while making sure conda dependencies work. Issue with making PATH not visible is that, dependencies of python modules like `_ssl` (and python extensions) are not found as they are in a different library `<prefix>/Library/bin`. To fix this I add `<prefix>/Library/bin` and other conda paths using `os.add_dll_directory` when the python interpreter is initialized. There are 2 unintended advantages with this,
1. Running python.exe and importing python extensions without activating python environment works.
2. C:\Windows\System32 has the lowest priority. (Previously we needed to use `CONDA_DLL_SEARCH_MODIFICATION=1` to make System32 have the lowest priority in python 3.6-3.9. That has lots of other issues too.)

In order to keep the 1st advantage, I had to keep the vendored zlib module as python on windows doesn't have a separate zlib extension, but is embedded into the interpreter as a builtin module. We can either,
1. Keep zlib vendored in
2. Copy `<prefix>/Library/bin/zlib.dll` to `<prefix>` in zlib conda package (85 KB file).
I'm fine with either option.

@conda-forge/python, @conda-forge/core, @chenghlee, @jezdez, @katietz, thoughts on the change of DLL search order to not look in PATH and also on how to proceed with zlib?

Reply:
Regarding `zlib`, I would go with either 1 or 2. If we go for 2, we should do the copy as part of the `zlib` package. Personally, I have a slight preference for 2 but I'm happy with 1 and we can do 2 as a follow-up PR.

Reply:
Re `zlib`, putting `zlib.dll` in the root for `conda-forge::zlib` isn't necessarily sufficient since we cannot assume users to have "clean" environments, i.e., it can happen that they have `defaults::zlib` installed (or install it) (I know this is not something we want, but esp. packages like `python` are installed by many unaware/inexperienced users). Less likely (probably not highly likely due to `zlib` having few dependencies), but still possible, is also that people get an older `zlib` build installed since `conda` does not prevent this.

Reply:
@mbargull, with https://github.com/conda-forge/zlib-feedstock/pull/52, this shouldn't be a problem as the new package `libzlib` will always have `zlib.dll` in the root. (Assuming defaults will use our recipe to have `libzlib`.)

Reply:
Hehe, yep, nice opportunity to split ;).